File: apps/frontend/lib/api/job-intake.ts
import { apiFetch, apiPost } from './client';

export type JobSourceType =
  | 'manual_text'
  | 'job_url'
  | 'pdf_url'
  | 'pdf_upload'
  | 'recruiter_message';

export type ExtractionMethod = 'manual' | 'deterministic' | 'http' | 'playwright' | 'pdf' | 'llm';

export interface DetectedJobLink {
  url: string;
  label?: string | null;
}

export interface ScreeningQuestion {
  id: string;
  question: string;
}

export interface DraftAnswer {
  question_id: string;
  answer: string;
  evidence: string[];
  needs_user_input: boolean;
  prompt: string;
}

export interface JobIntakeMetadata {
  source_type: JobSourceType;
  source_url?: string | null;
  source_title?: string | null;
  links: DetectedJobLink[];
  screening_questions: ScreeningQuestion[];
  draft_answers: DraftAnswer[];
  extraction_method: ExtractionMethod;
  warnings: string[];
  confidence: number;
}

export interface JobIntakeExtractRequest {
  source_type: Exclude<JobSourceType, 'pdf_upload'>;
  source_text?: string;
  url?: string;
  resume_id?: string | null;
  resume_text?: string | null;
}

export interface JobIntakeExtractResponse extends JobIntakeMetadata {
  job_description: string;
  raw_text: string;
  requires_review: boolean;
}

export interface JobIntakeConfirmRequest {
  job_description: string;
  resume_id?: string | null;
  intake_metadata?: JobIntakeMetadata | null;
}

export interface JobIntakeConfirmResponse {
  message: string;
  job_id: string;
  request: Record<string, unknown>;
}

async function parseJsonResponse<T>(response: Response, fallbackMessage: string): Promise<T> {
  let data: unknown;
  try {
    data = await response.json();
  } catch {
    throw new Error(fallbackMessage);
  }

  if (!response.ok) {
    const detail =
      data && typeof data === 'object' && 'detail' in data ? String(data.detail) : fallbackMessage;
    throw new Error(detail);
  }
  return data as T;
}

export async function extractJobIntake(
  request: JobIntakeExtractRequest
): Promise<JobIntakeExtractResponse> {
  const response = await apiPost('/jobs/intake/extract', request, 60_000);
  return parseJsonResponse<JobIntakeExtractResponse>(
    response,
    'Failed to extract the job description.'
  );
}

export async function uploadJobIntakePdf(
  file: File,
  resumeId?: string | null
): Promise<JobIntakeExtractResponse> {
  const formData = new FormData();
  formData.append('file', file);
  if (resumeId) {
    formData.append('resume_id', resumeId);
  }
  const response = await apiFetch(
    '/jobs/intake/pdf-upload',
    {
      method: 'POST',
      body: formData,
    },
    60_000
  );
  return parseJsonResponse<JobIntakeExtractResponse>(
    response,
    'Failed to extract the PDF job description.'
  );
}

export async function confirmJobIntake(
  request: JobIntakeConfirmRequest
): Promise<JobIntakeConfirmResponse> {
  const response = await apiPost('/jobs/intake/confirm', request);
  return parseJsonResponse<JobIntakeConfirmResponse>(
    response,
    'Failed to save the job description.'
  );
}

File: apps/backend/app/schemas/job_intake.py
"""Schemas for job-description intake automation."""

from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator, model_validator

JobSourceType = Literal[
    "manual_text",
    "job_url",
    "pdf_url",
    "pdf_upload",
    "recruiter_message",
]

ExtractionMethod = Literal[
    "manual",
    "deterministic",
    "http",
    "playwright",
    "pdf",
    "llm",
]


class DetectedJobLink(BaseModel):
    """Link detected during job intake."""

    url: str
    label: str | None = None


class ScreeningQuestion(BaseModel):
    """Recruiter screening question extracted separately from the JD."""

    id: str
    question: str


class DraftAnswer(BaseModel):
    """Evidence-only draft answer for a screening question."""

    question_id: str
    answer: str = ""
    evidence: list[str] = Field(default_factory=list)
    needs_user_input: bool = True
    prompt: str = ""


class JobIntakeMetadata(BaseModel):
    """Reviewed intake metadata persisted with the job record."""

    source_type: JobSourceType
    source_url: str | None = None
    source_title: str | None = None
    links: list[DetectedJobLink] = Field(default_factory=list)
    screening_questions: list[ScreeningQuestion] = Field(default_factory=list)
    draft_answers: list[DraftAnswer] = Field(default_factory=list)
    extraction_method: ExtractionMethod = "deterministic"
    warnings: list[str] = Field(default_factory=list)
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)


class JobIntakeExtractRequest(BaseModel):
    """Request to extract a JD from text or URL input."""

    source_type: Literal["manual_text", "job_url", "pdf_url", "recruiter_message"]
    source_text: str | None = None
    url: str | None = None
    resume_id: str | None = None
    resume_text: str | None = None

    @field_validator("source_text", "url", "resume_id", "resume_text", mode="before")
    @classmethod
    def _blank_to_none(cls, value: Any) -> Any:
        if isinstance(value, str) and not value.strip():
            return None
        return value

    @model_validator(mode="after")
    def _validate_source_payload(self) -> "JobIntakeExtractRequest":
        if self.source_type in ("manual_text", "recruiter_message") and not self.source_text:
            raise ValueError("source_text is required for this source type")
        if self.source_type in ("job_url", "pdf_url") and not self.url:
            raise ValueError("url is required for this source type")
        return self


class JobIntakeExtractResponse(BaseModel):
    """Extraction response shown in the review step."""

    source_type: JobSourceType
    job_description: str
    source_url: str | None = None
    source_title: str | None = None
    links: list[DetectedJobLink] = Field(default_factory=list)
    screening_questions: list[ScreeningQuestion] = Field(default_factory=list)
    draft_answers: list[DraftAnswer] = Field(default_factory=list)
    extraction_method: ExtractionMethod
    warnings: list[str] = Field(default_factory=list)
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    requires_review: bool = True


class JobIntakeConfirmRequest(BaseModel):
    """Persist a user-reviewed JD and optional intake metadata."""

    job_description: str
    resume_id: str | None = None
    intake_metadata: JobIntakeMetadata | None = None

    @field_validator("job_description")
    @classmethod
    def _validate_job_description(cls, value: str) -> str:
        normalized = value.strip()
        if len(normalized) < 50:
            raise ValueError("Job description is too short")
        return normalized


class JobIntakeConfirmResponse(BaseModel):
    """Response after reviewed intake is stored as a job."""

    message: str
    job_id: str
    request: dict[str, Any]
